src/main/kotlin/org/pkl/lsp/VirtualFile.kt

```diff
@@ -75,19 +79,58 @@ interface VirtualFile : ModificationTracker {
   /** Tells if this file is within a package. */
   val `package`: PackageDependency?
 
+  var contents: String
+
   fun parent(): VirtualFile?
 
   fun resolve(path: String): VirtualFile?
 
-  fun toModule(): PklModule?
-
-  fun contents(): String
+  fun getModule(): CompletableFuture<PklModule?>
 }
 
 sealed class BaseFile : VirtualFile {
-  internal val modificationCount: AtomicLong = AtomicLong(0)
+  override fun getModificationCount(): Long = version ?: -1L
+
+  override var version: Long? = null
+
+  abstract fun doReadContents(): String
 
-  override fun getModificationCount(): Long = modificationCount.get()
+  // Use if contents have not yet been set, read contents from external source, possibly causing
+  // I/O.
+  override var contents: String
+    get() {
+      return myContents ?: doReadContents()
+    }
+    set(text) {
+      myContents = text
+    }
+
+  final override fun getModule(): CompletableFuture<PklModule?> {
+    return project.cachedValuesManager.getCachedValue("VirtualFile($uri)") {
+      val future = CompletableFuture.supplyAsync { doBuildModule() }
+      CachedValue(future, this)
+    }!!
+  }
+
+  protected val logger by lazy { project.getLogger(this::class) }
+
+  private var myContents: String? = null
+
+  private val parser = Parser()
+
+  private fun doBuildModule(): PklModule? {
+    return try {
+      logger.log("building $uri")
+      val moduleCtx = parser.parseModule(contents)
+      return PklModuleImpl(moduleCtx, this)
+    } catch (e: LexParseException) {
+      logger.error("Parser Error building $file: ${e.message}")
+      null
```

Comment:
Note: previously, this was building parser diagnostics and sending them back to the client.

I was considering keeping the same logic, but `LexParseException` is going away once we swap out the parser.